tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,769 @@
+package tensorflow
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"reflect"
+	"runtime"
+	"unsafe"
+
+	"github.com/golang/protobuf/proto"
+
+	pb "github.com/tensorflow/tensorflow/tensorflow/contrib/go/proto"
+)
+
+import "C"
+
+const (
+	cAckByte  = 6
+	cBellByte = 7
+	cDc1      = 17
+
+	cBytesComplex64 = 8
+	cBytesFloat32   = 4
+	cBytesFloat64   = 8
+	cBytesInt16     = 2
+	cBytesInt32     = 4
+	cBytesInt64     = 8
+	cBytesUint16    = 2
+)
+
+// A DataType represents the type of the data contained in a Tensor
+type DataType pb.DataType
+
+// A TensorInterface is the interface implemented by Tensors.
+type TensorInterface interface {
```

Comment:
(Just to bring in history, when I looked at this, we had previously discussed "Tensorable" or something like that to follow the naming convention, but it seemed awkward.  The typical convention is to name interfaces "-er", and I like the idea of a Tensor being a concrete type.  Tensorer doesn't make sense grammatically.  I didn't have any brilliant ideas for this one.)

Given all that, perhaps we should just delete it.  After all, we don't have multiple competing Tensor implementations yet, so perhaps a Tensor is just a Tensor. :)

I suspect that what we'll find with experience is that *either* the user really wants a Tensor, or they want something with a subset of the Tensor methods, and will want a duck-typing interface for just that subset.

My vote is thus for getting rid of the interface for now.